00_Base/src/services/credentials.service.ts

```diff
@@ -474,6 +478,109 @@ export class CredentialsService {
+
+    // Make sure we stored the necessary version and version endpoints
+    // so that MSP can retrieve them later in the registration process
+    const storedVersion = await this.versionRepository.readOnlyOneByQuery(
+      {
+        where: {
+          version: versionNumber,
+        },
+        include: [VersionEndpoint],
+      },
+      OcpiNamespace.Version,
+    );
+    if (!storedVersion) {
+      throw new NotFoundError(`Version not found for version ${versionNumber}`);
+    }
+    const credentialsUrl = storedVersion.endpoints.find(
+      (endpoint: any) => endpoint.identifier === ModuleId.Credentials,
+    );
+    if (!credentialsUrl) {
+      throw new NotFoundError(
+        `Endpoint URL for ${ModuleId.Credentials} not found for version ${versionNumber}`,
+      );
+    }
+
+    const cpoTenantId = await this.getCpoTenantIdForServerRolesForRegistration(
+      serverCredentialsRoleDTOs,
+    );
+    const storedServerCredentialsRoles: ServerCredentialsRole[] = [];
+    if (!cpoTenantId) {
+      const cpoTenant = await CpoTenant.create(
+        {
+          serverCredentialsRoles: serverCredentialsRoleDTOs,
+        },
+        {
+          include: [
+            {
+              model: ServerCredentialsRole,
+              include: [
+                {
+                  model: BusinessDetails,
+                  include: [Image],
+                },
+              ],
+            },
+          ],
+        },
+      );
+      this.logger.info(`Created CpoTenant: ${JSON.stringify(cpoTenant)}`);
+      storedServerCredentialsRoles.push(...cpoTenant.serverCredentialsRoles);
+    } else {
+      storedServerCredentialsRoles.push(
+        ...(await this.serverCredentialsRoleRepository.createOrUpdateServerCredentialsRoles(
+          serverCredentialsRoleDTOs,
+          cpoTenantId,
+        )),
+      );
+    }
+
+    const credentialsTokenA = uuidv4();
+    // clientToken, clientCredentialsRoles and clientVersionDetails
+    // are added in the following post credentials requests based on the registration process
+    const clientInfo = await ClientInformation.create(
+      {
+        cpoTenantId: cpoTenantId,
+        registered: false,
+        serverToken: credentialsTokenA,
+        serverVersionDetails: [
+          {
+            version: versionNumber,
+            url: credentialsRequest.url,
+            endpoints: storedVersion.endpoints.map((endpoint) => ({
+              identifier: endpoint.identifier,
+              role: endpoint.role,
+              url: endpoint.url,
+            })),
+          },
+        ],
+      },
+      {
+        include: [
+          {
+            model: ServerVersion,
+            include: [Endpoint],
+          },
+        ],
+      },
+    );
+
+    const storedServerCredentialsRoleDTOs = storedServerCredentialsRoles.map(
+      (storedRole) => storedRole.toCredentialsRoleDTO(),
+    );
+    return CredentialsDTO.build(
```

Comment:
I see that in this method you have validated that we support the version, got existing tenant or created one if there wasnt one existing, then added server credentials with provided CredentialRole information and created the new ClientInformation with the newly generated `serverToken` which can be used by the sender for auth, and finally you return the CredentialsDTO in the response. All looks good - good job 👍

Reply:
it may be helpful to separate some of these chunks of code into smaller functions